src/main/scala/edu/cmu/cs/obsidian/typecheck/Checker.scala

```diff
@@ -2127,17 +2127,71 @@ private def checkStatement(
             logError(contract, NoConstructorError(contract.name))
         }
 
-        // TODO: enable multiple constructors. by adding appropriate dispatch logic in init. For now, we only support one constructor per contract.
-        if (constructors.length > 1) {
+        if (contract.isMain && constructors.length > 1) {
             logError(contract, MultipleConstructorsError(contract.name))
         }
 
-        val constructorsByArgTypes = constructors.groupBy(c => c.args.map(_.typIn.topPermissionType))
-        val matchingConstructors = constructorsByArgTypes.filter(_._2.size > 1)
+        def baseTypeName(arg: VariableDeclWithSpec): String =
+            arg.typIn match {
+                case t: NonPrimitiveType => t.contractName
+                case t => t.toString
+            }
+
+        // We have two group by's because the groupings are not exactly the same: this one disregards states
+        // to ensure that the generated Java code (which will not have states) will work
+        val constructorGroups = constructors.groupBy(_.args.map(baseTypeName))
+
+        def states(t: NonPrimitiveType): Set[String]=
+            t match {
+                case ContractReferenceType(contractType, permission, isRemote) =>
+                    table.lookupContract(contractType.contractName).map(_.possibleStates).getOrElse(Set())
+                case InterfaceContractType(name, simpleType) => states(simpleType)
+                case StateType(contractName, stateNames, isRemote) => stateNames
+            }
+
+        // Two types are distinguishable iff they are both owned and have non-overlapping state lists
+        def distinguishable(args: (ObsidianType, ObsidianType)): Option[String] = {
```

Comment:
The name and comment suggest that this should return a Boolean, but it does not. Please update the name and comment accordingly.